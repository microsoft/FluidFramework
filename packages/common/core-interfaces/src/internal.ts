/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

// eslint-disable-next-line no-restricted-syntax
export * from "./index.js";

// Important: all other exports must be type only exports. In package.json exports,
//  index.js is listed as the runtime file. This is done so that all imports are
//  using the same outer runtime file. (Could be changed if needed.)

// Export set of utility types re-tagged as internal for FF client convenience.
// These types are not intended for direct use by customers and api-extractor will
// flag misuse. If an externally visible version of these types is needed, import
// from via /internal/exposedUtilityTypes rather than /internal.
import type { DeepReadonly as ExposedDeepReadonly } from "./deepReadonly.js";
import type { InternalUtilityTypes as ExposedInternalUtilityTypes } from "./exposedInternalUtilityTypes.js";
import type {
	JsonDeserialized as ExposedJsonDeserialized,
	JsonDeserializedOptions,
} from "./jsonDeserialized.js";
import type {
	JsonSerializable as ExposedJsonSerializable,
	JsonSerializableOptions,
} from "./jsonSerializable.js";
import type {
	JsonTypeWith as ExposedJsonTypeWith,
	ReadonlyNonNullJsonObjectWith as ExposedReadonlyNonNullJsonObjectWith,
} from "./jsonType.js";

// Note: There are no docs for these re-exports. `@inheritdoc` cannot be used as:
//   1. api-extractor does not support renames.
//   2. api-extractor does not support package paths. ("Import paths are not supported")
// Also not useful, at least in VS Code, as substitution is not made in place.

/**
 * @internal
 */
export type DeepReadonly<T> = ExposedDeepReadonly<T>;

/**
 * @internal
 */
export type JsonDeserialized<
	T,
	Options extends JsonDeserializedOptions = {
		AllowExactly: [];
		AllowExtensionOf: never;
	},
> = ExposedJsonDeserialized<T, Options>;

/**
 * @internal
 */
export type JsonSerializable<
	T,
	Options extends JsonSerializableOptions = {
		AllowExactly: [];
		AllowExtensionOf: never;
	},
> = ExposedJsonSerializable<T, Options>;

/**
 * @internal
 */
export type JsonTypeWith<T> = ExposedJsonTypeWith<T>;

/**
 * @internal
 */
export type ReadonlyNonNullJsonObjectWith<T> = ExposedReadonlyNonNullJsonObjectWith<T>;

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace InternalUtilityTypes {
	/* eslint-disable jsdoc/require-jsdoc */
	export type FlattenIntersection<T extends ExposedInternalUtilityTypes.AnyRecord> =
		ExposedInternalUtilityTypes.FlattenIntersection<T>;
	export type IfSameType<
		X,
		Y,
		IfSame = unknown,
		IfDifferent = never,
	> = ExposedInternalUtilityTypes.IfSameType<X, Y, IfSame, IfDifferent>;
	/* eslint-enable jsdoc/require-jsdoc */
}
