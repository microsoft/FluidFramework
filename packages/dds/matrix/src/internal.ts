/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

// Privately exposed to MatrixUndoProvider
export const undoRemoveRows = Symbol("SharedMatrix.undoRemoveRows");
export const undoRemoveCols = Symbol("SharedMatrix.undoRemoveCols");
