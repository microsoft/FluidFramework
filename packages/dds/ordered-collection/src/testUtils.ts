/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { ConsensusResult, type IConsensusOrderedCollection } from "./interfaces.js";

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 * @internal
 */
export async function acquireAndComplete<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T | undefined> {
	let res: T | undefined;
	await collection.acquire(async (value: T) => {
		res = value;
		return ConsensusResult.Complete;
	});
	return res;
}

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 * @internal
 */
export async function waitAcquireAndComplete<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T> {
	let res: T | undefined;
	await collection.waitAndAcquire(async (value: T) => {
		res = value;
		return ConsensusResult.Complete;
	});
	return res as T;
}

/**
 * Helper method to acquire and release an item
 * Should be used in test code only
 * @internal
 */
export async function acquireAndRelease<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T | undefined> {
	let res: T | undefined;
	await collection.acquire(async (value: T) => {
		res = value;
		return ConsensusResult.Release;
	});
	return res;
}

/**
 * Helper method to acquire and release an item
 * Should be used in test code only
 * @internal
 */
export async function waitAcquireAndRelease<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T> {
	let res: T | undefined;
	await collection.waitAndAcquire(async (value: T) => {
		res = value;
		return ConsensusResult.Release;
	});
	return res as T;
}
