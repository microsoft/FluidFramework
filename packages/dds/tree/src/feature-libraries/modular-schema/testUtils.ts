/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { assert } from "@fluidframework/common-utils";
import { Delta } from "../../tree";

export interface MockChildChange {
    intentions: number[];
    ref: number;
}

export function mockChildChangeRebaser(
    change: MockChildChange,
    baseChange: MockChildChange,
): MockChildChange {
    assert(change.ref === baseChange.ref, "Invalid input passed to child rebaser");
    return {
        intentions: change.intentions,
        ref: baseChange.intentions[baseChange.intentions.length - 1],
    };
}

export function mockChildChangeInverter(change: MockChildChange): MockChildChange {
    return {
        intentions: change.intentions.map((i) => -i).reverse(),
        ref: change.intentions[change.intentions.length - 1],
    };
}

export function mockChildChangeComposer(changes: MockChildChange[]): MockChildChange {
    if (changes.length <= 1) {
        return changes[0];
    }
    const id: number[] = [];
    changes.forEach((change, i) => {
        const prev = changes[i - 1].intentions;
        assert(i === 0 || change.ref === prev[prev.length - 1], "Invalid input to child composer");
        id.push(...change.intentions);
    });
    return {
        intentions: id,
        ref: changes[0].ref,
    };
}

export function mockChildChangeToDelta(change: MockChildChange): Delta.Modify {
    return { type: Delta.MarkType.Modify, setValue: { value: change.intentions.map(String).join("|") } };
}
