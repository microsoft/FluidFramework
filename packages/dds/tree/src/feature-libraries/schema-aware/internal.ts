/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

// Used by put public types, but not part of the desired API surface
export {
	TypeSetToTypedTrees as TreeTypesToTypedTreeTypes,
	TypedSchemaData,
	TypedTree,
	CollectOptions,
	TypedFields,
	ApplyMultiplicity,
	ValueFieldTreeFromSchema,
	FlexibleObject,
} from "./schemaAware";

export { NamesFromSchema, ValuesOf, TypedValue, PrimitiveValueSchema } from "./schemaAwareUtil";
