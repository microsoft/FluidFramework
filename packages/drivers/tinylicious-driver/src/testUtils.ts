/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

export { createInsecureTinyliciousTestTokenProvider } from "./insecureTinyliciousTokenProvider.js";

export {
	createInsecureTinyliciousTestUrlResolver,
	createTinyliciousTestCreateNewRequest,
} from "./insecureTinyliciousUrlResolver.js";
