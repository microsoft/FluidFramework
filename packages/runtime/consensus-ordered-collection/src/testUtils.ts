/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import {
    ConsensusResult,
    IConsensusOrderedCollection,
} from "./interfaces";

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 */
export async function acquireAndComplete<T>(collection: IConsensusOrderedCollection<T>): Promise<T | undefined> {
    let res: T | undefined;
    await collection.acquire(async (value: T) => {
        res = value;
        return ConsensusResult.Complete;
    });
    return res;
}

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 */
export async function waitAcquireAndComplete<T>(collection: IConsensusOrderedCollection<T>): Promise<T> {
    let res: T | undefined;
    await collection.waitAndAcquire(async (value: T) => {
        res = value;
        return ConsensusResult.Complete;
    });
    return res as T;
}
