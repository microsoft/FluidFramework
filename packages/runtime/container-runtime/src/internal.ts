/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

/* Exports for internal use only
 * This is only used during transition of legacy+alpha code to internal.
 * Here there are conflicts between index reexports with deprecations and clean internal versions.
 * In cases where * exports conflict with named exports, the named exports take precedence per
 * https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#sec-getexportednames.
 * This does trigger the `import/export` lint warning (which is intentionally disabled here). This
 * approach ensures that a non-deprecated version of ContainerRuntime (provided as named direct
 * export) eclipses the deprecated one from `./index.ts`.
 * Alternatively, another file could be created for external exports adding ContainerRuntime and
 * they'd share common exports files (the rest of current index.ts).
 */
/* eslint-disable import/export */
/* eslint-disable import/no-deprecated */

// eslint-disable-next-line no-restricted-syntax
export * from "./index.js";
import { ContainerRuntime as ContainerRuntimeClass } from "./containerRuntime.js";
export type ContainerRuntime = ContainerRuntimeClass;
export const ContainerRuntime = ContainerRuntimeClass;
