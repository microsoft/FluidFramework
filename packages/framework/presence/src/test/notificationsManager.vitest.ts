/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { EventAndErrorTrackingLogger } from "@fluidframework/test-utils/internal";
import { useFakeTimers, type SinonFakeTimers } from "sinon";
import { describe, it, afterAll, afterEach, beforeAll, beforeEach, expect } from "vitest";

import { Notifications, type PresenceNotifications } from "../index.js";
import type { createPresenceManager } from "../presenceManager.js";

import { MockRuntimeSignalSnapshotter } from "./snapshotEphemeralRuntime.js";
import { prepareConnectedPresence } from "./testUtils.js";

describe("Presence", () => {
	describe("NotificationsManager", () => {
		describe("snapshot tests", () => {
			// IMPORTANT: All tests in this suite have an extra signal for each workspace that is initialized in the test.
			// This is a bug. See AB#24392. This means that when looking at snapshots from these tests, the "real" snapshots
			// start at 2, not 1, for most tests. Some tests may have additional incorrect leading signals. Such examples
			// are noted inline.
			let runtime: MockRuntimeSignalSnapshotter;
			let logger: EventAndErrorTrackingLogger;
			const initialTime = 1000;
			let clock: SinonFakeTimers;
			let presence: ReturnType<typeof createPresenceManager>;

			beforeAll(async () => {
				clock = useFakeTimers();
			});

			beforeEach(() => {
				logger = new EventAndErrorTrackingLogger();
				runtime = new MockRuntimeSignalSnapshotter(logger);

				// We are configuring the runtime to be in a connected state, so ensure it looks connected
				runtime.connected = true;

				// Note that while the initialTime is set to 1000, the prepareConnectedPresence call advances
				// it to 1010 so all tests start at that time.
				clock.setSystemTime(initialTime);

				// Disable submitting signals with a dummy function. This ensures we don't capture signals from
				// test setup, like the prepareConnectedPresence call.
				const submitSignalOriginal = runtime.submitSignal;
				runtime.submitSignal = () => {};

				// Set up the presence connection
				presence = prepareConnectedPresence(runtime, "sessionId-2", "client2", clock, logger);

				// Restore the submiSignal function
				runtime.submitSignal = submitSignalOriginal;
			});

			afterEach(() => {
				clock.reset();
			});

			afterAll(() => {
				clock.restore();
			});

			it("notification signals are sent immediately", async () => {
				// Configure a notifications workspaces
				// eslint-disable-next-line @typescript-eslint/ban-types
				const notificationsWorkspace: PresenceNotifications<{}> = presence.getNotifications(
					"name:testNotificationWorkspace",
					{},
				);

				notificationsWorkspace.add(
					"testEvents",
					Notifications<
						// Below explicit generic specification should not be required.
						{
							newId: (id: number) => void;
						},
						"testEvents"
					>(
						// A default handler is not required
						{
							newId: (client, newId) => {
								// do nothing?
							},
						},
					),
				);

				const { testEvents } = notificationsWorkspace.props;

				clock.tick(10); // Time is now 1020

				clock.tick(30); // Time is now 1050
				// SIGNAL #1
				testEvents.emit.broadcast("newId", 77);

				clock.tick(10); // Time is now 1060
				// SIGNAL #2
				testEvents.emit.broadcast("newId", 88);

				expect(runtime.submittedSignals).toHaveLength(2);

				// Verify first signal set the newId to 77
				let signal = runtime.submittedSignals[0];
				expect(
					// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
					(signal?.[1] as any).data["n:name:testNotificationWorkspace"].testEvents[
						"sessionId-2"
					].value.args,
				).toEqual([77]);

				// Verify first signal set the newId to 88
				signal = runtime.submittedSignals[1];
				expect(
					// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
					(signal?.[1] as any).data["n:name:testNotificationWorkspace"].testEvents[
						"sessionId-2"
					].value.args,
				).toEqual([88]);
			});
		});
	});
});
