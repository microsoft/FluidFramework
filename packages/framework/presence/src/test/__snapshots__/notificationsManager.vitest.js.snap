// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Presence > NotificationsManager > snapshot tests > notification signals are sent immediately > submitted signal 1`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1050,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "n:name:testNotificationWorkspace": {
        "testEvents": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 0,
            "value": {
              "name": "newId",
              "args": [
                77
              ]
            },
            "ignoreUnmonitored": true
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > NotificationsManager > snapshot tests > notification signals are sent immediately > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1060,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "n:name:testNotificationWorkspace": {
        "testEvents": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 0,
            "value": {
              "name": "newId",
              "args": [
                88
              ]
            },
            "ignoreUnmonitored": true
          }
        }
      }
    }
  }
]"
`;
